### PR TITLE
Add Secaggplus example

### PR DESCRIPTION
**Pull request by panh99**
### Description

We don't have an example for using the SecAgg+ protocol in Flower

### Related issues/PRs

This PR is based on the following PR:
Add Secure Aggregation functionalities #2159

### Explanation

Add `examples/secagg-mt-pytorch` example

### Checklist

- [x] Implement proposed change
- [x] Write tests
- [ ] Update [documentation](https://flower.dev/docs/writing-documentation.html)
- [ ] Update [changelog](https://github.com/adap/flower/blob/main/doc/source/changelog.rst)
- [x] Make CI checks pass
- [x] Ping maintainers on [Slack](https://flower.dev/join-slack/) (channel `#contributions`)

### CODE REVIEW

File: examples/secagg-mt-pytorch/run_clients.sh [hunk 1]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+
+# Number of client processes to start
+N=5 # Replace with your desired value
+
+# Start N client processes
+for i in $(seq 1 $N)
+do
+  nohup python client.py &
```
Comment (danieljanes) [inline examples/secagg-mt-pytorch/run_clients.sh:9]: Is `nohup` required here?

File: examples/secagg-mt-pytorch/client.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+
+# Define Flower client with the SecAgg/SecAgg+ protocol
+class FlowerClient(fl.client.NumPyClient, SecAggPlusHandler):
+    def fit(self, parameters, config):
+        # Force a significant delay for teshing purposes
+        if self._shared_state.sid == 0:
+            time.sleep(40)
+        if IS_VALIDATION:
+            return [np.zeros(10000)], 1, {}
```
Comment (danieljanes) [inline examples/secagg-mt-pytorch/client.py:30]: Is this part still required?

File: examples/secagg-mt-pytorch/run_clients.sh (hunk 1 above)
Comment (panh99) [reply examples/secagg-mt-pytorch/run_clients.sh]: no, not really. just wanted to hide output from `client.py`. I've removed nohup

File: examples/secagg-mt-pytorch/client.py (hunk 2 above)
Comment (panh99) [reply examples/secagg-mt-pytorch/client.py]: no. I've removed it.

File: examples/secagg-mt-pytorch/task.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+import torch.nn.functional as F
+from torch.utils.data import DataLoader
+from torchvision.datasets import CIFAR10
+from torchvision.transforms import Compose, Normalize, ToTensor
+from tqdm import tqdm
+
+warnings.filterwarnings("ignore", category=UserWarning)
+DEVICE = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+NUM_ITERATIONS = 3
+IS_VALIDATION = True
```
[the remark this answers is not in the document]
Comment (panh99) [reply examples/secagg-mt-pytorch/task.py]: no. I've removed it.

File: examples/secaggplus-mt/driver.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+                task,
+                task_pb2.Task(
+                    producer=node_pb2.Node(
+                        node_id=0,
+                        anonymous=True,
+                    ),
+                    consumer=node_pb2.Node(
+                        node_id=sampled_node_id,
+                        anonymous=False,
+                        # Must be False for this Secure Aggregation example
```
Comment (danieljanes) [inline examples/secaggplus-mt/driver.py:38]:
```suggestion
                        # Must be False for this Secure Aggregation example
                        anonymous=False,
```

File: examples/secaggplus-mt/settings.py [hunk 5]
```diff
@@ -0,0 +1,3 @@
+LOG_EXPLAIN = True
+
+NUM_ROUNDS = 3
```
Comment (danieljanes) [inline examples/secaggplus-mt/settings.py:3]: We can probably hard-code these in the respective modules and get rid of this one, just to simplify things a bit

File: examples/secaggplus-mt/workflows.py [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+        key_dict = _get_from_task(task)
+        pk1, pk2 = key_dict[KEY_PUBLIC_KEY_1], key_dict[KEY_PUBLIC_KEY_2]
+        sid2public_keys[nid2sid[node_id]] = [pk1, pk2]
+
+    """
+    =============== Share keys stage ===============   
+    """
+    if LOG_EXPLAIN:
+        print(f"\nForwarding public keys...")
+    # Braodcast public keys to clients
```
Comment (danieljanes) [inline examples/secaggplus-mt/workflows.py:188]:
```suggestion
    # Broadcast public keys to clients
```

File: examples/secaggplus-mt/workflows.py [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+        client_masked_vec = named_values[KEY_MASKED_PARAMETERS]
+        client_masked_vec = [bytes_to_ndarray(b) for b in client_masked_vec]
+        if LOG_EXPLAIN:
+            print(f"Received {client_masked_vec[1]} from Client {nid2sid[node_id]}.")
+        masked_vector = parameters_addition(masked_vector, client_masked_vec)
+    masked_vector = parameters_mod(masked_vector, mod_range)
+    """
+    =============== Unmask stage ===============   
+    """
+    if LOG_EXPLAIN:
```
Comment (danieljanes) [inline examples/secaggplus-mt/workflows.py:274]:
```suggestion
    masked_vector = parameters_mod(masked_vector, mod_range)

    """
    =============== Unmask stage ===============   
    """

    if LOG_EXPLAIN:
```

File: examples/secaggplus-mt/workflows.py [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+    if len(surviving_node_ids) < threshold:
+        raise Exception("Not enough available clients after unmask vectors stage")
+    for _, task in node_messages.items():
+        named_values = _get_from_task(task)
+        for owner_sid, share in zip(
+            named_values[KEY_SECURE_ID_LIST], named_values[KEY_SHARE_LIST]
+        ):
+            collected_shares_dict[owner_sid].append(share)
+    # Remove mask for every client who is available before ask vectors stage,
+    # Divide vector by first element
```
Comment (danieljanes) [inline examples/secaggplus-mt/workflows.py:306]:
```suggestion
    # divide vector by first element
```

**Review by danieljanes — CHANGES_REQUESTED** [review]

File: examples/secaggplus-mt/settings.py (hunk 5 above)
Comment (panh99) [reply examples/secaggplus-mt/settings.py:3]: sure, fixed it.

File: examples/secaggplus-mt/client.py [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+            print(f"Client {self._shared_state.sid} dropped for testing purposes.")
+            time.sleep(4)
+            return [np.ones(3)], 1, {}
+        ret = [np.ones(3)]
+        print(f"Client {self._shared_state.sid} uploading {ret[0]}")
+        return ret, 1, {}
+
+
+# Start Flower client
+fl.client.start_client(
```
[the remark this answers is not in the document]
Comment (panh99) [reply examples/secaggplus-mt/client.py:31]: I've updated the client class. The SecAgg+ handler does not work with numpy client. `start_numpy_client()` will wrap the client class and hence cause `handle()` in `message_handler.py` cannot detect the `client.handle_secure_aggregation()` handler.

[the remark this answers is not in the document]
Comment (panh99) [reply examples/secaggplus-mt/client.py:31]: Will fix this problem later. I just updated the client class to let it inherit from `Client` instead.

File: examples/secaggplus-mt/client.py [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+class FlowerClient(fl.client.Client, SecAggPlusHandler):
+    def fit(self, fit_ins: FitIns) -> FitRes:
+        ret_vec = [np.ones(3)]
+        ret = FitRes(
+            status=Status(code=Code.OK, message="Success"),
+            parameters=ndarrays_to_parameters(ret_vec),
+            num_examples=1,
+            metrics={},
+        )
+        # Force a significant delay for teshing purposes
```
Comment (danieljanes) [inline examples/secaggplus-mt/client.py:21]:
```suggestion
        # Force a significant delay for testing purposes
```

**Review by danieljanes — APPROVED** [review]
Lgtm!

**Review by danieljanes — APPROVED** [review]

**Pull request merged**